### ISSUE TITLE
Improvement Request: option after trailing args?

### ISSUE THREAD

**Issue by FurcyPin** [scallop/scallop#147]
Hi there,

First, I would like to thank the creator(s) of this library, it is quite awesome and I have been
using it for a few years for my own project now:  https://github.com/flaminem/flamy

Several users often asked me if it could be possible to add options 
at the end of the command line, _i.e._ after the trailing args.

Here is a simple example of what I mean:

```
import org.rogach.scallop._

class Conf(arguments: Seq[String]) extends ScallopConf(arguments) {
  val apples = opt[Int](required = true)
  val bananas = opt[Int]()
  val name = trailArg[String]()
  verify()
}

object ScallopDemo extends App {

/* 
 * this works: 
 */
//  val conf = new Conf("--apples 4 --bananas 10 strangeTree".split(" "))

  
/* 
 * This returns an error: 
 * [scallop] Error: Required option 'name' not found
*/ 
  val conf = new Conf("--bananas 10 strangeTree --apples 4".split(" "))

  println("tree name is: " + conf.name())
  println("apples are: " + conf.apples())
  println("bananas are: " + conf.bananas())

}
```

Do you think it would be possible to make the second string work ? 
I can't think of a reason why not, as many projects with a CLI allow it, but perhaps 
there is I reason why Scallop cannot do it.

P.S. : If you are interested, for my project I even made a interactive shell with autocomplete for command, options and even their arguments. Now that it is done, thanks to scallop, when I add a new command or option to my project, I get auto-completion for free ! 
I even use custom ValueConverters to get autocomplete for the arguments.
You can see some asciicast demonstrating the project [here](http://flamy.readthedocs.io/en/latest/Demo.html)

The code behind it is not very clean at the moment, and is too specific for my project to be used for another, but if you want to have a look, you can try [this class](https://github.com/flaminem/flamy/blob/master/src/main/scala/com/flaminem/flamy/exec/shell/CliArgs.scala) or [this one](https://github.com/flaminem/flamy/blob/master/src/main/scala/com/flaminem/flamy/exec/shell/OptionCompleter.scala).
 
Also, beware that I use an old version of Scallop (0.9.5) at the moment, I need to upgrade it.

If anyone is interested to do something similar for his/her project, I would be glad to discuss about it, and perhaps I could even try making a generic library for others to use.

Have a nice day

**Comment by Rogach** [comment]
Hi! Flamy looks nice, as well as auto-generated autocompletion.

Regarding trailing args - there might be some confusion with error reporting. For example:

```
--one-arg-option option1 typo --other-option trailing_arg
```

How should the error look in this case? "Trailing args parse error: typo trailing_arg" seems quite confusing for me.

**Comment by FurcyPin** [comment]
I guess we should at least have a rule that says that you can't have an option between trailing args,
only [options] [trailing_args] [more options].
That way, I guess the error could be something like 
`Bad arguments for option 'one-arg-option': 'option1 typo'`.

It might be complicated to do, though, because the parsing of `--one-arg-option option1 typo` would depend
on the presence of a trailing_arg at the end.

**Comment by Rogach** [comment]
If we want to stay close to often-used utils, then restricting trailing args to appear only once does not work. For example, my version of grep does accept many sequences of trailing args:

```
grep -P "waltz" -o log.txt -n
```
The biggest source of problems is that most utils do not allow variable length lists of arguments to options, while Scallop does.

**Comment by FurcyPin** [comment]
Indeed, I was thinking about other kind of tools like ansible that only take one single trailing arg. I guess it is simpler in this case too.

**Comment by jpork** [comment]
Trailing args should not be followed by options otherwise it wouldn't be possible to pass options put into trailing args to embedded tools. 
Java does this for instance like so:

`java -cp *.jar DemoMain --opt1 arg1 --opt2 arg2`

see also  #203

**Comment by FurcyPin** [comment]
I guess this issue can be closed, then.

For what it's worth: while developing a similar tool to the one where I had this need, 
I ended up replacing the trailing args with a named option, so in the end it looks like that.

```
my_tool --opt1 opt1-arg --trailing-args-name trailing-arg-1 trailing-arg-2 --opt2 opt2-arg
```

It's a bit annoying to specify the `--trailing-args-name` every time but at the same time it's clearer for the users and we can add options after them, so I'm quite happy with that, actually.

**Comment by Rogach** [comment]
@FurcyPin I think this one is different - you were talking about options that are present in the definition, while #203 is about completely unrelated options that are meant to be passed along to some other programs, so Scallop won't be able to distinguish correct options from mistyped ones. 

#203 can be solved with `--` argument separator (everything after `--` is passed into trailing args without any changes). Your case is different, though.

I still hope to implement this, but I am very short on time, so I can't give any timeline.

**Comment by rcannood** [comment]
I'd just like to mention that I would also be very interested in finding a solution to this issue.

**Comment by Rogach** [comment]
I published a snapshot with this feature:

```
resolvers += "Sonatype OSS Snapshots" at "https://oss.sonatype.org/content/repositories/snapshots"
libraryDependencies += "org.rogach" %% "scallop" % "4.0.0-SNAPSHOT"
```

**Comment by rcannood** [comment]
Works great for me!

**Comment by Rogach** [comment]
Released in version `4.0.0`.
